### PR TITLE
add example: <Generate Embeddings> and <Embedding Similarity Search>

### PR DESCRIPTION
**Pull request by aceld**

### CODE REVIEW

**Comment by sashabaranov** [comment]
Hey, thank you for this PR!

**Comment by aceld** [comment]
@sashabaranov  Hello, I believe that this example is what many developers need. It is a basic case of using Embedding for semantic search. I hope you can review and approve it. Thank you.

**Comment by sashabaranov** [comment]
@aceld could you please make changes based on the comments above? Would love to merge this PR after the changes are made

**Comment by aceld** [comment]
@sashabaranov  Which comment do you want me to edit?

File: README.md [hunk 1]
```diff
@@ -462,3 +462,229 @@ func main() {
...
+// Calculate cosine similarity
+func cosineSimilarity(v1, v2 []float32) float32 {
+	dot := dotProduct(v1, v2)
+	v1Magnitude := math.Sqrt(float64(dotProduct(v1, v1)))
+	v2Magnitude := math.Sqrt(float64(dotProduct(v2, v2)))
+	return float32(float64(dot) / (v1Magnitude * v2Magnitude))
+}
+
+// Calculate dot product
+func dotProduct(v1, v2 []float32) float32 {
```
[the remark this answers is not in the document]
Comment (sashabaranov) [reply README.md:593]: @aceld

File: README.md [hunk 2]
```diff
@@ -462,3 +462,229 @@ func main() {
...
+		return nil, err
+	}
+
+	return resp.Data[0].Embedding, nil
+}
+
+// Calculate cosine similarity
+func cosineSimilarity(v1, v2 []float32) float32 {
+	dot := dotProduct(v1, v2)
+	v1Magnitude := math.Sqrt(float64(dotProduct(v1, v1)))
```
[the remark this answers is not in the document]
Comment (sashabaranov) [reply README.md:587]: @aceld

**Review by sashabaranov — CHANGES_REQUESTED** [review]

**Comment by sashabaranov** [comment]
Duh, I think comments were in draft stage and not published, sorry!

**Comment by aceld** [comment]
@sashabaranov  done.

File: embeddings_utils.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// We recommend cosine similarity. The choice of distance function typically doesn’t matter much.
+// OpenAI embeddings are normalized to length 1, which means that:
+// Cosine similarity can be computed slightly faster using just a dot product.
+// Cosine similarity and Euclidean distance will result in the identical rankings.
+// See: https://platform.openai.com/docs/guides/embeddings/limitations-risks
+func CosineSimilarity(v1, v2 []float32) float32 {
+	// Calculate dot product
+	dot := DotProduct(v1, v2)
+	// Calculate magnitude of v1
+	v1Magnitude := math.Sqrt(float64(DotProduct(v1, v1)))
```
Comment (sashabaranov) [inline embeddings_utils.go:16]: Embeddings are normalized to length 1, so we don't need to do that. CosineSimilarity is equal to DotProduct in this case.

https://platform.openai.com/docs/guides/embeddings/which-distance-function-should-i-use

File: embeddings_test.go [hunk 4]
```diff
@@ -89,3 +89,41 @@ func TestEmbeddingEndpoint(t *testing.T) {
...
+		t.Errorf("Unexpected result. Expected: %v, but got %v", expected, result)
+	}
+}
+
+func TestDotProduct(t *testing.T) {
+	v1 := []float32{1, 2, 3}
+	v2 := []float32{2, 4, 6}
+	expected := float32(28.0)
+	result := DotProduct(v1, v2)
+	if result != expected {
```
Comment (sashabaranov) [inline embeddings_test.go:118]: You can't compare floats like that https://bitbashing.io/comparing-floats.html

File: README.md [hunk 5]
```diff
@@ -462,3 +462,214 @@ func main() {
 }
 ```
 </details>
+
+<details>
+<summary>Generate Embeddings</summary>
```
Comment (sashabaranov) [inline README.md:467]: There are a couple of problems with these examples:

1. They are too specific, containing non-relevant parts of file reading, binary decoding, and sorting
2. I would prefer to keep README neutral, without specific mentions of developers and their preferences
3. Both examples are not that different from being included together

I suggest we include a single minimal example that:
1. Takes a bunch of strings
2. Computes their embeddings
3. Finds the one closest to a query

And let's use fine food reviews [dataset](https://www.kaggle.com/datasets/snap/amazon-fine-food-reviews) that OpenAI has in their [docs](https://platform.openai.com/docs/guides/embeddings/use-cases):

```go
examples := []string{
	"It was tasty and fresh. The other one I bought was old and tasted moldy. But this one was good.",
	"Great coffee at a good price. I'm a subscription buyer and I buy this month after month. What more can I say?",
	"This chocolate is amazing..I love the taste and smell, this is the only chocolate for me...I found a new love!"
	"I love this coffee!  And such a great price.  Will buy more when I am running out which will be soon.",
	"The Raspberry Tea Syrup is great. I can use it for hot and cold drinks as well in certain recipes.",
	"Everyone that dips with this loves it!  So easy to use! Olive oil and tasty bread is all you need.",
	"This is a favorite of mine for using over ice. Even bought it to give out as Christmas gifts last year.",
	"If you like a great , hot, sauce then buy this. If spicy with heat isn't to your liking then don't buy it.",
}

query := "chocolate"
```

File: embeddings_utils.go (hunk 3 above)
Comment (aceld) [reply embeddings_utils.go:16]: @sashabaranov  OK I just delete the function CosineSimilarity and it’s over. Anyway, this is not used in the example.

File: embeddings_test.go (hunk 4 above)
Comment (aceld) [reply embeddings_test.go:118]: @sashabaranov  So how to compare? Can you provide some case parameters?

File: README.md (hunk 5 above)
Comment (aceld) [reply README.md:467]: @sashabaranov  The purpose of providing binary storage is to simulate the vector database scenario. If it is only compared by []string, it does not conform to the normal business process.

File: embeddings_test.go (hunk 4 above)
Comment (sashabaranov) [reply embeddings_test.go:98]: @aceld something like

```go
func isClose(a, b float32) bool {
        if a == b {
		return true
	}
	return math.Abs(float64(a-b)) < 1e-12
}
```

https://floating-point-gui.de/errors/comparison/

File: README.md (hunk 5 above)
Comment (sashabaranov) [reply README.md:467]: @aceld I think computing embeddings for all `examples []string` would be sufficient for the README example.

**Comment by sashabaranov** [comment]
@aceld if you have some time — maybe let's remove all README changes, fix float comparison and merge!

**Comment by aceld** [comment]
OK @sashabaranov

**Comment by aceld** [comment]
@sashabaranov  I apologize for not being able to find time to make the changes recently. I am sorry for any inconvenience caused, and I will try my best to allocate time next week to fix this issue.

**Comment by aceld** [comment]
@sashabaranov 
I have completed the fix according to your requirements, but the Codecov Report did not pass. I don't quite understand it. Please enlighten me. Thank you.

**Comment by sashabaranov** [comment]
@aceId hey, you imports are not sorted properly. Please refer to https://golangci-lint.run/ for more information

**Comment by aceld** [comment]
> @aceid hey, you imports are not sorted properly. Please refer to https://golangci-lint.run/ for more information

@sashabaranov done!

File: embeddings_utils.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package openai
+
+// DotProduct Calculate dot product of two vectors.
+func DotProduct(v1, v2 []float32) float32 {
```
Comment (sashabaranov) [inline embeddings_utils.go:4]: Why not put it as an Embedding method?

File: README.md [hunk 7]
```diff
@@ -516,29 +516,225 @@ func main() {
...
+		return
+	}
+	defer file.Close()
+
+	encoder := gob.NewEncoder(file)
+	err = encoder.Encode(selectionsEmbeddings)
+	if err != nil {
+		fmt.Printf("Encode error: %v\n", err)
+		return
+	}
```
Comment (sashabaranov) [inline README.md:594]: I think file I/O and marshalling is largely out of scope to the purpose of this example. Could you please remove it?

File: README.md [hunk 8]
```diff
@@ -516,29 +516,225 @@ func main() {
...
+	for _, embed := range allEmbeddings {
+		// OpenAI embeddings are normalized to length 1, which means that:
+		// Cosine similarity can be computed slightly faster using just a dot product
+		score := openai.DotProduct(embed, inputEmbd)
+		questionScores = append(questionScores, score)
+	}
+
+	// Take the subscripts of the top few selections with the highest similarity
+	sortedIndexes := sortIndexes(questionScores)
+	sortedIndexes = sortedIndexes[:3] // Top 3
```
Comment (sashabaranov) [inline README.md:684]: Could we please add this section to the previous example and have one single example for embeddings?

**Review by sashabaranov — CHANGES_REQUESTED** [review]

File: embeddings_utils.go (hunk 6 above)
Comment (aceld) [reply embeddings_utils.go:4]: image

@sashabaranov  like you say.

File: README.md (hunk 7 above)
Comment (aceld) [reply README.md:594]: @sashabaranov  Sure, you're right. Do you have any suggestions on how to store vector data more efficiently? I would appreciate some advice.

Comment (sashabaranov) [reply README.md:594]: @aceld I think the storage of vector data is largely is out of scope for this README — the point is just to show an example, not to build vector-search DB.

File: embeddings_utils.go (hunk 6 above)
Comment (sashabaranov) [reply embeddings_utils.go:4]: Totally, but we don't really have `[]float32` vectors in the library except for `Embedding` struct. Might make sense to add it as `func (e Embedding) DotProduct(another Embedding)`

**Comment by Leeaandrob** [comment]
It's fantastic how can I help this pr to be merged? @aceld @sashabaranov ?

**Comment by sashabaranov** [comment]
@Leeaandrob yes, please feel free to fork it and continue in another PR!

**Comment by aceld** [comment]
@sashabaranov @Leeaandrob The code conflicts have been resolved, and currently, all checks have passed.

**Comment by sashabaranov** [comment]
@aceld there are still a number of changes to be made mentioned above. To re-iterate:

- We don't need two huge examples (see https://github.com/sashabaranov/go-openai/pull/274/files#r1232058195) — Please make one small example without `encoding/gob` etc.
- I'm not sure this change is relevant here https://github.com/sashabaranov/go-openai/pull/274/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L596
- Let's make `DotProduct` a method of `Embedding` struct (see https://github.com/sashabaranov/go-openai/pull/274/files#r1232242888)

**Pull request closed without merging**